From 5b29f69a38ce25dfe1706b4481c4d608d71a0006 Mon Sep 17 00:00:00 2001
From: Jin Li <jinl@codeaurora.org>
Date: Thu, 13 Oct 2016 17:26:49 -0400
Subject: [PATCH 2/2] compositor-drm: add proper stride/handle for NV12 UV
 plane

When call AddFB2 for YUV format plane, the correct stride and bo
handle need to be passed into API. Otherwise, kernel driver won't
be able to parse it.

Change-Id: I3a27a0841aab07a1d1833f16aa34156ac4312783
---
 src/compositor-drm.c | 25 ++++++++++++++++++++++++-
 1 file changed, 24 insertions(+), 1 deletion(-)

diff --git a/src/compositor-drm.c b/src/compositor-drm.c
index af5bed6..c9ae01f 100644
--- a/src/compositor-drm.c
+++ b/src/compositor-drm.c
@@ -360,6 +360,17 @@ static struct gl_renderer_interface *gl_renderer;
 
 static const char default_seat[] = "seat0";
 
+#define is_pseudo_planar_format(f) \
+	(((f) == GBM_FORMAT_NV12) || ((f) == GBM_FORMAT_NV21) || \
+	 ((f) == GBM_FORMAT_NV16) || ((f) == GBM_FORMAT_NV61))
+
+#define is_planar_format(f) \
+	(((f) == GBM_FORMAT_YUV410) || ((f) == GBM_FORMAT_YVU410) || \
+	 ((f) == GBM_FORMAT_YUV411) || ((f) == GBM_FORMAT_YVU411) || \
+	 ((f) == GBM_FORMAT_YUV420) || ((f) == GBM_FORMAT_YVU420) || \
+	 ((f) == GBM_FORMAT_YUV422) || ((f) == GBM_FORMAT_YVU422) || \
+	 ((f) == GBM_FORMAT_YUV444) || ((f) == GBM_FORMAT_YVU444))
+
 /**
  * Return a string describing the type of a DRM object
  */
@@ -1134,7 +1145,19 @@ drm_fb_get_from_bo(struct gbm_bo *bo,
 	if (format && !backend->no_addfb2) {
 		handles[0] = fb->handle;
 		pitches[0] = fb->stride;
-		offsets[0] = 0;
+		offsets[0] = gbm_bo_get_offset(bo);
+
+		/* Need to plug in with bufdef utility once it's ready */
+		switch (format) {
+		case GBM_FORMAT_NV12:
+			handles[1] = fb->handle;
+			pitches[1] = fb->stride;
+			offsets[1] = offsets[0] + ((width + 127) & ~127) * ((height + 31) & ~31);
+			break;
+		default:
+			if (is_pseudo_planar_format(format) || is_planar_format(format))
+				weston_log("Unsupported format 0x%x", format);
+		}
 
 		ret = drmModeAddFB2(backend->drm.fd, width, height,
 				    format, handles, pitches, offsets,
-- 
2.10.1

